### PR TITLE
Feature/grading instructions/sgi feedback reference

### PR DESCRIPTION
**Pull request by HanyaElhashemy**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

We can derive lots of useful data and statistics about the performance of the students from the assessment of their submissions. Beside the final total score we can also derive more data from each submission element and their assessment which resembles a partial assessment. A link between the partial assessment and the applied grading instruction would therefore be very beneficial. 
Through this link instructors can conclude which part of the assignment the students were struggling with or which part the student performed especially good at.
In addition, this link allows us to handle subsequent mistakes (Folge Fehler)
### Description

We save for each feedback element the id of the grading instruction applied on this feedback of a submission element.
We check the maxCountInScore value set by the instructors which resembles how often points for a single instruction should be added to the total score if applied multiple times.
If the number of application is higher than the number of maxCountInScore then the points are not added to the final score.
The default value 0 of maxCountInScore resemble that there is no limit of application, meaning that the instruction can be applied as often as possible.
### Changes to the Database:
In order to tie a grading instruction to a feedback element I added the column `grading_instruction_id` which saves the id of the grading instruction applied to the feedback element. One Grading instruction could be applied to many feedback elements therefore I added a one to many relation between grading_instruction and feedback table. (I followed here the same principle as the one implemented between Participation and Submission: one Participation <-> many submissions)
Another approach would be to save a physical reference string, like the one that refers the feedback to the text segment (in text exercises) or SVG (in modeling exercises). I didn't choose this approach because I thought it might be restrictive in the future if I want to automatically update all submitted assessments when the grading instructions are updated. 
image

--TODO--
adjust apollon to support this feature for modeling exercise assessment as well
### Steps for Testing

1.login as an instructor and create an file upload/text exercise with structured grading criteria. Make sure to sett at least for one instruction the maxCountInScore attribute with a value higher than 0
2. login as a student and submit solutions for these exercises you just created
3. login as a tutor and asses the students submissions
4. open the result after submit or override and check that the score is calculated and displayed correctly and subsequent mistakes are taken into consideration
5. check in general (for modeling and programming ex) that the score calculation still works correctly for all other exercise types

### Screenshots

image
image

### CODE REVIEW

**Comment by HanyaElhashemy** [comment]
@tessaRuckstuhl Thanks so much for your review ^^.  This is an aspect I did not think about so thank you for shedding the light on it, I did implement it now. So now it is disabled to change the score for grading instructions where the maxCountInScore limit is set. If the maxCountInScore is 0 then you are allowed to edit the score as you wish. In both cases it is still positive to edit the feedback text or drag&drop a different SGI

Regarding the file upload exercises: the file upload exercises do not support feedback elements with 0P. This was an implementation decision taken some time ago and is not part of this PR scope.

**Comment by HanyaElhashemy** [comment]
@tessaRuckstuhl same logic is applied for text exercises now as well :)

File: src/main/webapp/app/entities/feedback.model.ts [hunk 1]
```diff
@@ -19,6 +20,7 @@ export enum FeedbackType {
 
 export class Feedback implements BaseEntity {
     public id: number;
+    public gradingInstruction: GradingInstruction;
```
Comment (krusche) [inline src/main/webapp/app/entities/feedback.model.ts:23]: I guess this can be null, right?

**Comment by krusche** [comment]
@HanyaElhashemy thanks for the contribution! Could you please merge the conflict in master.xml
In addition, I think we should have some server and client integration tests to test whether usage count and the reference are working correctly.

In particular the logic in the services should be tested, including edge cases!

File: src/main/webapp/app/entities/feedback.model.ts (hunk 1 above)
Comment (HanyaElhashemy) [reply src/main/webapp/app/entities/feedback.model.ts:23]: yes

**Comment by Alexanderkrp** [comment]
Good work 👍

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 2]
```diff
@@ -176,7 +177,39 @@ private double calculateTotalScore(Double calculatedScore, Double maxScore) {
      * @param assessments the List of Feedback
      * @return the total score
      */
-    protected Double calculateTotalScore(List<Feedback> assessments) {
-        return assessments.stream().mapToDouble(Feedback::getCredits).sum();
+    public static Double calculateTotalScore(List<Feedback> assessments) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:180]: This should not be a static method.

File: src/main/java/de/tum/in/www1/artemis/domain/Result.java [hunk 3]
```diff
@@ -39,6 +34,8 @@
 @JsonInclude(JsonInclude.Include.NON_EMPTY)
 public class Result implements Serializable {
 
+    // private static AssessmentService assessmentService;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/Result.java:37]: please remove this change

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 4]
```diff
@@ -176,7 +178,39 @@ private double calculateTotalScore(Double calculatedScore, Double maxScore) {
      * @param assessments the List of Feedback
      * @return the total score
      */
-    protected Double calculateTotalScore(List<Feedback> assessments) {
-        return assessments.stream().mapToDouble(Feedback::getCredits).sum();
+    public Double calculateTotalScore(List<Feedback> assessments) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:181]: you changed the method completely :-) please also adapt the JavaDoc comment above.
Some additional inline comments would also help

File: src/test/javascript/spec/service/structured-grading-criterion.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    let service: StructuredGradingCriterionService;
+    let httpMock: HttpTestingController;
+    let expectedResult: any;
+    let feedbacks: Feedback[];
+
+    beforeEach(() => {
+        TestBed.configureTestingModule({
+            imports: [HttpClientTestingModule],
+        });
+        expectedResult = {} as HttpResponse<ExerciseHint>;
```
Comment (krusche) [inline src/test/javascript/spec/service/structured-grading-criterion.service.spec.ts:23]: please fix the warning `'expectedResult' is assigned a value but never used`

File: src/test/javascript/spec/service/structured-grading-criterion.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        feedbacks.push(feedback1);
+        feedbacks.push(feedback2);
+        feedbacks.push(feedback3);
+        feedbacks.push(feedback4);
+        feedbacks.push(feedback5);
+        feedbacks.push(feedback6);
+    });
+
+    describe('Service methods', () => {
+        it('should calculate the total score', async () => {
```
Comment (krusche) [inline src/test/javascript/spec/service/structured-grading-criterion.service.spec.ts:71]: can you add a second example of a test with some other values?
The idea is that each case in the new method in structured-grading-criterion.service.ts is executed at least once!

Comment (HanyaElhashemy) [reply src/test/javascript/spec/service/structured-grading-criterion.service.spec.ts:26]: I added a 2nd example with negative and decimal numbers.
All different types of feedbacks are covered as well!

**Pull request merged**